api/v1alpha1/amaltheasession_children.go

```diff
@@ -0,0 +1,220 @@
+package v1alpha1
+
+import (
+	"context"
+	"fmt"
+	"reflect"
+
+	appsv1 "k8s.io/api/apps/v1"
+	v1 "k8s.io/api/core/v1"
+	networkingv1 "k8s.io/api/networking/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/apimachinery/pkg/util/intstr"
+	"k8s.io/utils/pointer"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+// StatefulSet returns a AmaltheaSession StatefulSet object
+func (cr *AmaltheaSession) StatefulSet() appsv1.StatefulSet {
+	labels := labelsForAmaltheaSession(cr.Name)
+	replicas := int32(1)
+	if cr.Spec.Hibernated {
+		replicas = 0
+	}
+
+	session := cr.Spec.Session
+	pvc := cr.PVC()
+
+	sessionContainer := v1.Container{
+		Image:           session.Image,
+		Name:            "session",
+		ImagePullPolicy: v1.PullIfNotPresent,
+
+		Ports: []v1.ContainerPort{{
+			ContainerPort: session.Port,
+			Name:          "session-port",
+			Protocol:      v1.ProtocolTCP,
+		}},
+
+		Args:                     session.Args,
+		Command:                  session.Command,
+		Env:                      session.Env,
+		Resources:                session.Resources,
+		VolumeMounts:             []v1.VolumeMount{{Name: "session", MountPath: session.Storage.MountPath}},
+		TerminationMessagePath:   "/dev/termination-log",
+		TerminationMessagePolicy: v1.TerminationMessageReadFile,
+	}
+
+	securityContext := &v1.SecurityContext{
+		RunAsNonRoot: &[]bool{true}[0],
+		RunAsUser:    &[]int64{session.RunAsUser}[0],
+		RunAsGroup:   &[]int64{session.RunAsGroup}[0],
+	}
+
+	if session.RunAsUser == 0 {
+		securityContext.RunAsNonRoot = &[]bool{false}[0]
+	}
+
+	sessionContainer.SecurityContext = securityContext
+
+	containers := []v1.Container{sessionContainer}
+	containers = append(containers, cr.Spec.ExtraContainers...)
+
+	sts := appsv1.StatefulSet{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:            cr.Name,
+			Namespace:       cr.Namespace,
+			OwnerReferences: []metav1.OwnerReference{cr.OwnerReference()},
+		},
+		Spec: appsv1.StatefulSetSpec{
+			Replicas: &replicas,
+			Selector: &metav1.LabelSelector{
+				MatchLabels: labels,
+			},
+			Template: v1.PodTemplateSpec{
+				ObjectMeta: metav1.ObjectMeta{
+					Labels: labels,
+				},
+				Spec: v1.PodSpec{
+					Containers:     containers,
+					InitContainers: cr.Spec.ExtraInitContainers,
+					Volumes:        []v1.Volume{{Name: "session", VolumeSource: v1.VolumeSource{PersistentVolumeClaim: &v1.PersistentVolumeClaimVolumeSource{ClaimName: pvc.Name}}}},
```

Comment:
Two things here:
- Again, the question of constant for the Name field
- Formatting wise, it's the only place that packs the structure on one line

Also, are we sure that we will always have a PVC ? Thinking about the case of anonymous session and also the ephemeral use cases (BinderHub like).